modules/@shopify/checkout-sheet-kit/src/index.ts

```diff
@@ -78,63 +81,125 @@ class ShopifyCheckoutSheet implements ShopifyCheckoutSheetKit {
     event: CheckoutEvent,
     callback: CheckoutEventCallback,
   ): EmitterSubscription | undefined {
-    if (event === 'pixel' && typeof callback === 'function') {
-      const eventHandler = callback as (data?: PixelEvent) => void;
-
-      /**
-       * Event data can be sent back as either a parsed Pixel Event object or a JSON string.
-       */
-      const cb = (eventData: string | PixelEvent) => {
-        try {
-          if (typeof eventData === 'string') {
-            try {
-              const parsed = JSON.parse(eventData);
-
-              if (
-                parsed.hasOwnProperty('customData') &&
-                typeof parsed.customData === 'string'
-              ) {
-                try {
-                  parsed.customData = JSON.parse(parsed.customData);
-                } catch {}
-              }
-              eventHandler(parsed as PixelEvent);
-            } catch (error) {
-              const parseError = new WebPixelsParseError(
-                'Failed to parse Web Pixel event data: Invalid JSON',
-                {
-                  cause: 'Invalid JSON',
-                },
-              );
-              // eslint-disable-next-line no-console
-              console.error(parseError);
-            }
-          } else if (eventData && typeof eventData === 'object') {
-            eventHandler(eventData);
-          }
-        } catch (error) {
-          const parseError = new WebPixelsParseError(
-            'Failed to parse Web Pixel event data',
-            {
-              cause: 'Unknown',
-            },
-          );
-          // eslint-disable-next-line no-console
-          console.error(parseError);
-        }
-      };
-
-      // Web Pixel event specific handler
-      return ShopifyCheckoutSheet.eventEmitter.addListener(event, cb);
+    let eventCallback: CheckoutEventCallback = callback;
+
+    switch (event) {
+      case 'pixel':
+        eventCallback = this.handleWebPixelEvent(
+          callback,
+        ) as PixelEventCallback;
+        break;
+      case 'completed':
+        eventCallback = this.handleCompletedEvent(
+          callback,
+        ) as CheckoutCompletedEventCallback;
+        break;
+      default:
+        eventCallback = callback;
     }
 
     // Default handler for all non-pixel events
-    return ShopifyCheckoutSheet.eventEmitter.addListener(event, callback);
+    return ShopifyCheckoutSheet.eventEmitter.addListener(event, eventCallback);
   }
 
   public removeEventListeners(event: CheckoutEvent) {
     ShopifyCheckoutSheet.eventEmitter.removeAllListeners(event);
   }
+
+  // ---
+
+  private handleWebPixelEvent(
+    callback: CheckoutEventCallback,
+  ): (eventData: string | PixelEvent) => void {
```

Comment:
The return type is also confusing. The function that it's returned is always casted to `PixelEventCallback`. That's either wrong or this function could be simplified.

Reply:
Oh hang on, think I got it to work

Reply:
What I mean is that we can just update the type of `PixelEventCallback` to what reflects the code:

`export type PixelEventCallback = (eventData: string | PixelEvent) => void;`

Then we update the types on the function and remove the casting because the types will really reflect what's happening:

```ts
public addEventListener(
    event: CheckoutEvent,
    callback: CheckoutEventCallback,
  ): EmitterSubscription | undefined {
    let eventCallback: CheckoutEventCallback;

    switch (event) {
      case 'pixel':
        eventCallback = this.handleWebPixelEvent(
          callback,
        );
        break;
      case 'completed':
        eventCallback = this.handleCompletedEvent(
          callback,
        );
        break;
      default:
        eventCallback = callback;
    }

    // Default handler for all non-pixel events
    return ShopifyCheckoutSheet.eventEmitter.addListener(event, eventCallback);
  }

...

  private handleWebPixelEvent(
    callback: CheckoutEventCallback,
  ): PixelEventCallback {
    const eventHandler = callback as PixelEventCallback;
    return (eventData: string | PixelEvent): void => { }
      
  }
```

Reply:
Here's how I solved this: https://github.com/Shopify/checkout-sheet-kit-react-native/pull/76/commits/fd62c5f9b414d85a409869d98957009bd5adb17f

Reply:
> We can just update the type of PixelEventCallback to what reflects the code:

We can't, because then consumers would have to expect a `string | PixelEvent` value, and parse it themselves

Reply:
I was confusing with `callback` and `eventCallback` now. 

What I was initially referring to in this comment was this cast which was wrong and was fixed here: https://github.com/Shopify/checkout-sheet-kit-react-native/pull/76/commits/48013c7531d0eb05e1bbb3d7083b72ffb2f7665e


All good now!